Add Validated.andThen

ceedubs opened the pull request:
Fixes #604.

This is essentially `flatMap` for `Validated`, but since `flatMap`
suggests monadic bind in Cats and this is not consistent with `ap`,
another name is being used.

This PR is not complete yet. I need to add a Scaladoc comment and tut documentation about this change. However, I figured there might be some discussion about the name of the method, so I want to wait until a name is decided to start writing docs :). Is `andThen` too overloaded?

adelbertc commented:
`andThen` is fine by me, but it might confuse people seeing `andThen` from `FunctionN`. I know Scoobi used `mapFlatten`, or at least did at some point..

mpilquist commented:
:+1: I'm okay with `andThen`.

non commented:
I would be fine with `andThen`. In the context of `Validated` I could also imagine calling it something about *valid* e.g. `whenValid`, `mapValid`, etc.

ceedubs commented:
It sounds like people are generally okay with `andThen`, so I'll go with that. We can reconsider in the future if it's tripping people up.

I'll move forward with documentation.

ceedubs commented:
I added some documentation.

@adelbertc I deleted part of your `Validated` docs about `flatMap`, because it didn't seem to belong after the `andThen` docs, but I'd like your input on this change.

I think there's still room for improvement, so please feel free to chime in.

adelbertc commented:
LGTM :+1:
